V2

connor-compto opened the pull request.

connor-compto committed:
deploy to devnet

Bonsuano committed:
change comptoken proof to bitcoin compatible

Bonsuano committed:
mining pool functional

connor-compto committed:
fix linter warnings

connor-compto committed:
notes

connor-compto committed:
fix unit tests + simplify

connor-compto committed:
forgot

connor-compto committed:
determine COMPTOKEN_ACCOUNT_SPACE

connor-compto committed:
use @compto/comptoken.js and fix component tests

connor-compto committed:
fix integration test

connor-compto committed:
get devnet test working

connor-compto committed:
cleanup rust

connor-compto committed:
add world id accounts

connor-compto committed:
cpi into world id

connor-compto committed:
forgot

connor-compto committed:
repeat verification prevention - pda

connor-compto committed:
fix bugs

connor-compto committed:
fix bugs

connor-compto committed:
pass in min rent

connor-compto committed:
Merge pull request #17 from compto-com/repeat_verification_prevention_1_pda

repeat verification prevention - pda

connor-compto committed:
minor fixes

connor-compto committed:
Delete comptoken/notes.md

connor-compto committed:
fix verify_human account order comment

connor-compto committed:
fix ubi distribution bug + add test

connor-compto committed:
improve handling of instruction data

connor-compto committed:
improve handling of instruction data

connor-compto committed:
fix wsl in windows subsystem issue

connor-compto committed:
fix some issues from merging

connor-compto committed:
Merge branch 'master' into worldcoin-verification

connor-compto committed:
fix bug in devnet tests

connor-compto committed:
fix nullifier pda creation

connor-compto committed:
update tests, and fix bugs

connor-compto committed:
move difficulty_calculator.py from stratum server

connor-compto committed:
fix bugs w/ world accounts

connor-compto committed:
allow non-forceful account creation in py scripts

connor-compto committed:
get compto_public_keys from file-local cache

connor-compto committed:
use devnet compto_public_keys + cached test_user in devnet tests

connor-compto committed:
.

connor-compto committed:
ci: add github action to run tests on PR

connor-compto committed:
feat: print exception info on background process __exit__

connor-compto committed:
ci: flush ci test prints

connor-compto committed:
ci: split tests job into smaller jobs

connor-compto committed:
refactor: make seperate build script

connor-compto committed:
ci: create setup steps

connor-compto committed:
ci: fix create token cmd

connor-compto committed:
chore: update spl-token-2022 from 4.0.0 to v9.0.0

connor-compto committed:
chore: update spl-token-2022 from 4.0.0 to v9.0.0

connor-compto committed:
ci: add setup solana action

connor-compto committed:
ci: /dev/null solana-test-validator output

connor-compto committed:
ci: explicitly use localhost for solana-test-validator

connor-compto committed:
ci: actually upload .cache to artifacts

connor-compto committed:
ci: print validator logs on fail

connor-compto committed:
ci: start test-validator in build

connor-compto committed:
ci: add debug prints to test-validator startup

connor-compto committed:
ci: persist payer keypair

connor-compto committed:
ci: use actual address for mocked solana cmd files

connor-compto committed:
test: update getOwedUbiComptokens

connor-compto committed:
fixup! fix bugs w/ world accounts

connor-compto committed:
test: add worldId accounts

connor-compto committed:
test: update verifyAccount test

connor-compto committed:
test: verify human test

connor-compto committed:
feat: re-verification

connor-compto committed:
refactor: make instructions module and move ixs to it

also some renames and clean up imports

connor-compto committed:
refactor: use enum for account signer/writable

connor-compto committed:
refactor: make instruction data verification functions

connor-compto committed:
refactor: make account verification functions

connor-compto committed:
chore: remove old comments

connor-compto committed:
feat: add flag_stale_account instruction

connor-compto committed:
fix: minting requires account be current

connor-compto committed:
feat: add reverify_human instruction

connor-compto committed:
feat: ignore stale users/comptokens for daily distribution

connor-compto committed:
chore: delete unneeded round ties even

connor-compto committed:
feat: add unverify instructions

connor-compto committed:
refactor: move reverify_human instruction

connor-compto committed:
refactor: move global_data to data

connor-compto committed:
fix: incorrect account writable/signer checks

connor-compto committed:
test: update tests

connor-compto committed:
feat: "burn" stale ubi on collect

connor-compto committed:
feat: "burn" by redistributing as ubi

connor-compto committed:
chore: format

connor-compto committed:
refactor: rename tests to match program

connor-compto committed:
test: add reverifyHumanInstruction test

connor-compto committed:
fix: div by 0 in verify human

connor-compto committed:
doc: unverifyHuman ix data

connor-compto committed:
fix: prevent double dipping "future" ubi

connor-compto committed:
test: add unverifyHumanInstruction test

connor-compto committed:
test: unverifyHuman2Instruction

connor-compto committed:
fix: increment verified_humans in verify human

connor-compto committed:
refactor: stale -> inactive

connor-compto committed:
fix: bugs in flag_inactive_account and interest calc

connor-compto committed:
test: add flagInactiveAccountInstruction test

connor-compto committed:
fix: review things

connor-compto committed:
fix: check log_dir prop exists

connor-compto committed:
fix: remove --no-reset from component tests ivoke

connor-compto committed:
Merge branch 'master' into worldcoin-verification

connor-compto committed:
Merge branch 'add-auto-test' into worldcoin-verification

connor-compto committed:
fix: issue from merge

connor-compto committed:
test: print timeout output in component tests

connor-compto committed:
test: fix broken tests

connor-compto committed:
test: reduce output clutter

connor-compto committed:
Merge branch 'master' into worldcoin-verification

connor-compto committed:
feat: add solana_world_id_program binary

connor-compto committed:
fix: build fail b/c dep

connor-compto committed:
test: fix compute unit issue

connor-compto committed:
feat: allow collecting for others sometimes

connor-compto committed:
chore: update comptoken.js dep

connor-compto committed:
feat: init v2

connor-compto committed:
feat: init mints

connor-compto committed:
feat: add ring buffer

connor-compto committed:
feat: add global data daily distribution data

connor-compto committed:
refactor: split daily distribution data into own struct

connor-compto committed:
feat: add daily distribution methods

connor-compto committed:
feat: add valid_blockhashes + fix bugs

connor-compto committed:
feat: zero-copy global data

connor-compto committed:
refactor: rename some things

locked -> staked
unlocked -> unstaked
initialize::handler -> initialize::initialize
ring_buffer::position as usize -> ring_buffer::position()

connor-compto committed:
refactor: move hash to own file

connor-compto committed:
refactor: use bytemuck instead of manually bit-fiddling

connor-compto committed:
feat: add user data

connor-compto committed:
feat: add create user data instruction

connor-compto committed:
feat: add resize user data account instruction

connor-compto committed:
feat: add stake and unstake instructions

connor-compto committed:
test: update init test

connor-compto committed:
test: add create user data test

connor-compto committed:
refactor: rename some variables/constants

connor-compto committed:
test: add test mint unchecked ixs

connor-compto committed:
feat: add collect instruction

connor-compto committed:
feat: add daily distribution instruction

connor-compto committed:
feat: add get valid blockhashes instruction

connor-compto committed:
feat: add get submit mining proof instruction

connor-compto committed:
feat: custom errors

connor-compto committed:
feat: add verification instructions

connor-compto committed:
feat: make global data mint authority

connor-compto committed:
test: fix broken tests

connor-compto committed:
refactor: replace v1 with v2

connor-compto committed:
test: add bankrun dep to tests

connor-compto committed:
test: add collect test

connor-compto committed:
fix: account mutability issues

connor-compto committed:
fix: early adopter ubi bugs

connor-compto committed:
test: improve collect test

connor-compto committed:
test: create test utils folder, move utils there

connor-compto committed:
fixup! test: improve collect test

connor-compto committed:
fixup! test: create test utils folder, move utils there

connor-compto committed:
chore: add logs for yield/ubi calc

connor-compto committed:
test: add more collect tests

connor-compto committed:
fix: bugs in collect mut/signers & default yield rate

connor-compto committed:
test: add another collect test

connor-compto committed:
test: fix user/global data account serialization

connor-compto committed:
test: split collect tests into own file

connor-compto committed:
test: add more collect tests + stubs

connor-compto committed:
test: move initialize to own file

connor-compto committed:
test: split initialize into smaller tests

connor-compto committed:
test: move createUserDataAccount to own file + add stubs

connor-compto committed:
fix: bugs in dailyDistribution

connor-compto committed:
test: add daily distribution tests

connor-compto committed:
test: group daily distribution tests into sub-suites

connor-compto committed:
test: add last daily distribution tests

connor-compto committed:
test: cleanup daily distribuion tests

connor-compto committed:
test: add tests for core getValidBlockhashes

connor-compto committed:
test: add core resizeUserDataAcount tests

connor-compto committed:
test: add stake/unstake core tests + fix bugs

connor-compto committed:
fix: bugs in submit mining proof

connor-compto committed:
test: core submitMiningProof tests

connor-compto committed:
test: more submitMiningProof tests

connor-compto committed:
test: add script to find test proofs

connor-compto committed:
tests: add stubs for verification ixs

Bonsuano commented on programs/comptoken/src/instructions/verification/mod.rs:
```diff
@@ -0,0 +1,436 @@
+
+    // 5. update global data
+    global_data.daily_distribution.verified_accounts_count += 1;
+    global_data.daily_distribution.remaining_early_adopter_count -= 1;
```
these should only increment if `!nullifier_exists` correct?

connor-compto replied:
verified accounts count should increment regardless, but remaining early adopter count should only decrement if `!nullifier_exists`. this has been fixed

Bonsuano commented on idls/solana_world_id_program.json:
```diff
@@ -0,0 +1,1421 @@
+        {
+          "name": "guardian_signatures",
+          "writable": true,
+          "signer": true
```
PDAs cannot be signers

connor-compto replied:
idls are generated, and solana world id program was developed by wormhole, with testing and auditing. it should work.
I ended up including the actual program in this project with this [PR](https://github.com/compto-com/comptoken-program/pull/27), but I generally don't want to touch it beyond the bare minimum necessary to get it working with our program. a lot of the more inscrutable aspects of the idl are clear in the code (e.g. the const `0` on line 736 is an enum value for a specific verification  type)

Bonsuano commented on idls/solana_world_id_program.json:
```diff
@@ -0,0 +1,1421 @@
+              {
+                "kind": "const",
+                "value": [
+                  0
```
why is there a const `0` for a seed? Shouldn't bump seeds be runtime calculated?

connor-compto replied:
the full solana-world-id-program was pulled into this project, so the idl was removed. it includes more context, but the `[0]` isn't a bump seed it's an enum value for verification type

Bonsuano commented on idls/solana_world_id_program.json:
```diff
@@ -0,0 +1,1421 @@
+                  201,
+                  174,
+                  146,
+                  162
```
these seeds of random looking numbers need a `docs` explaining where they come from and what they mean

connor-compto replied:
the seeds are clearer in the src, which is now part of the program. that one is a specific program's id (bpf_loader_upgradeable)

Bonsuano commented on idls/solana_world_id_program.json:
```diff
@@ -0,0 +1,1421 @@
+      ],
+      "args": [
+        {
+          "name": "guardian_signatures",
```
its a bit confusing that `guardian_signatures` is the name of both an account and arg, consider a slight suffix like `_arg` or something to distinguish between them

connor-compto replied:
renamed acct to `guardian_signatures_buffer` and arg to `new_guardian_signatures`

Bonsuano commented on idls/solana_world_id_program.json:
```diff
@@ -0,0 +1,1421 @@
+          "writable": true
+        },
+        {
+          "name": "latest_root",
```
seems like `clean_up_root` might need `latest_root` to be writable?

connor-compto replied:
`clean_up_root` removes stale `root` accounts, unless it is also the latest root stored in the `latest_root` account.

Bonsuano commented on idls/solana_world_id_program.json:
```diff
@@ -0,0 +1,1421 @@
+  },
+  "instructions": [
+    {
+      "name": "claim_ownership",
```
in general, instructions could benefit from `docs` explanations of what they are intended to do

connor-compto replied:
added docs in #31

Bonsuano commented on programs/comptoken/src/constants.rs:
```diff
@@ -0,0 +1,64 @@
+#[constant]
+// the target end daily max increase. this value achieves ~25% max increase over the course of a year. this value was chosen by taking
+// the USD supply increase per year (~7%), and quadrupling it to allow for periods of larger growth, then rounding to a nicer number.
+pub const END_GOAL_PERCENT_INCREASE: f64 = 0.00061;
```
```suggestion
// TODO: consider moving from floating point to integer arithmetic
pub const END_GOAL_PERCENT_INCREASE: f64 = 0.00061;
```

connor-compto replied:
there is no (easy) integer equivalent for `powf`, and the floating point math works fine.

Bonsuano commented on programs/comptoken/src/constants.rs:
```diff
@@ -0,0 +1,64 @@
+// TODO: is there a better name for this?
+// this value was chosen by roughly simulating the distribution and choosing a value that reasonably
+// paces the distribution towards the end goal without being too aggressive
+pub const ADJUST_FACTOR: f64 = 0.3;
```
```suggestion
// TODO: consider moving from floating point to integer arithmetic
pub const ADJUST_FACTOR: f64 = 0.3;
```

connor-compto replied:
see [above](https://github.com/compto-com/comptoken-program/pull/26#discussion_r3104426001)

connor-compto committed:
Solana world id program (#27)

* test: add (broken) verify tests

* feat: add solana-world-id-program

* fix: bug in claim ownership

* test: add test-with-logs script

* test: fix verification  test

* fix: bugs in verify

* test: add reverification tests

* test: add tests for core unverify/unverify2

* test: add missing core verify test

* ci: update ci tests

* test: use comptoken.js in tests

* test: fix tests

* ci: fix test ci

* test: update comptoken.js

* docs: add comments

* Enforce mutual exclusivity of mainnet and testnet features

Added compile error for mutually exclusive features 'mainnet' and 'testnet'.

* Enforce mutual exclusivity of mainnet and testnet features

Added compile error for mutually exclusive features 'mainnet' and 'testnet'.

* chore: cleanup solana-world-id app

* script: fix-idl only search specific toml section

* test: GuardianSet deserialize buffer len check

* refactor: rename vars to include time unit

connor-compto committed:
Initialize script (#28)

* feat: add initialize script

* feat: update world app id

* feat: update solana-world-id app

* fix: problem requiring idl fix script

* test: update comptoken.js

* docs: improve code comments

* fix: [u8;1] works better for anchor

* fix: endianness mismatch issue

* fix: resize user data problem

* fix: review comments

* docs: add comment about userdata max size

* ci: remove call to deleted anchor script

* app: minor fixes

* feat: add verificaitonType const

* fix: incorrect verify values

* fix: query type verification

connor-compto committed:
Initialize script comments (#29)

* fix: minor bugs in app

* feat: add nBits constant

* refactor: consistent acct names

* fix: revert EARLY_ADOPTER_COUNT

* ci: use anchor/solana versions from toml

* ci: fix component tests

* refactor: {get->sync}ValidBlockhashes

* test: get->sync in tests

* ci: fix component tests

connor-compto committed:
Rename unverify (#30)

* refactor: more descriptive unverify names

* test: fix unverify tests

connor-compto committed:
fix: solana-world-id-program program id

connor-compto committed:
Last fixes (#31)

* refactor: rename guardian_signatures

* docs: add comments for ixs

* ci: fix ci testing

* ci: actually fix ci testing

connor-compto committed:
fix: mining reward amount (#32)
